### PR TITLE
Juniper EX4400 Series and ACX7100-32C

### PR DESCRIPTION
**Pull request by suom1**

### CODE REVIEW

File: device-types/Juniper/EX4400-24MP.yaml [hunk 1]
```diff
@@ -0,0 +1,77 @@
+---
+manufacturer: Juniper
+model: EX4400-24MP
+slug: juniper-ex4400-24mp
+part_number: EX4400-24MP
+u_height: 1
+is_full_depth: true
+airflow: front-to-rear
```
Comment (danner26) [inline device-types/Juniper/EX4400-24MP.yaml:8]: Please add `weight` if you have/can find that information

File: device-types/Juniper/EX4400-24T.yaml [hunk 2]
```diff
@@ -0,0 +1,76 @@
+---
+manufacturer: Juniper
+model: EX4400-24T
+slug: juniper-ex4400-24t
+part_number: EX4400-24T
+u_height: 1
+is_full_depth: true
```
Comment (danner26) [inline device-types/Juniper/EX4400-24T.yaml:7]: Please add `weight` and `airflow` if you have/can find that information

File: device-types/Juniper/EX4400-24X.yaml [hunk 3]
```diff
@@ -0,0 +1,76 @@
+---
+manufacturer: Juniper
+model: EX4400-24X
+slug: juniper-ex4400-24x
+part_number: EX4400-24X
+u_height: 1
+is_full_depth: true
```
Comment (danner26) [inline device-types/Juniper/EX4400-24X.yaml:7]: Please add `weight` and `airflow` if you have/can find that information

File: device-types/Juniper/EX4400-48F.yaml [hunk 4]
```diff
@@ -0,0 +1,124 @@
+---
+manufacturer: Juniper
+model: EX4400-48F
+slug: juniper-ex4400-48f
+part_number: EX4400-48F
+u_height: 1
+is_full_depth: true
```
Comment (danner26) [inline device-types/Juniper/EX4400-48F.yaml:7]: Please add `weight` and `airflow` if you have/can find that information

File: device-types/Juniper/EX4400-48MP.yaml [hunk 5]
```diff
@@ -0,0 +1,125 @@
+---
+manufacturer: Juniper
+model: EX4400-48MP
+slug: juniper-ex4400-48mp
+part_number: EX4400-48MP
+u_height: 1
+is_full_depth: true
+airflow: front-to-rear
```
Comment (danner26) [inline device-types/Juniper/EX4400-48MP.yaml:8]: Please add `weight` if you have/can find that information

File: device-types/Juniper/EX4400-48P.yaml [hunk 6]
```diff
@@ -0,0 +1,125 @@
+---
+manufacturer: Juniper
+model: EX4400-48P
+slug: juniper-ex4400-48p
+part_number: EX4400-48p
+u_height: 1
+is_full_depth: true
+airflow: front-to-rear
```
Comment (danner26) [inline device-types/Juniper/EX4400-48P.yaml:8]: Please add `weight` if you have/can find that information

File: device-types/Juniper/EX4400-48T.yaml [hunk 7]
```diff
@@ -0,0 +1,124 @@
+---
+manufacturer: Juniper
+model: EX4400-48T
+slug: juniper-ex4400-48t
+part_number: EX4400-48T
+u_height: 1
+is_full_depth: true
```
Comment (danner26) [inline device-types/Juniper/EX4400-48T.yaml:7]: Please add `weight` and `airflow` if you have/can find that information

File: device-types/Juniper/EX4400-48T.yaml [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+    type: 100gbase-x-qsfp28
+  - name: et-0/1/1
+    type: 100gbase-x-qsfp28
+console-ports:
+  - name: Console
+    type: rj-45
+  - name: Front Console
+    type: usb-c
+power-ports:
+  - name: PSU0
```
Comment (danner26) [inline device-types/Juniper/EX4400-48T.yaml:119]: It looks like this devices power supplies are modular. If this is the case, then the PSUs should actually be `module-bays` and the applicable PSU units should be created and added to the module-types folder

File: device-types/Juniper/EX4400-48P.yaml [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+    type: 100gbase-x-qsfp28
+  - name: et-0/1/1
+    type: 100gbase-x-qsfp28
+console-ports:
+  - name: Console
+    type: rj-45
+  - name: Front Console
+    type: usb-c
+power-ports:
+  - name: PSU0
```
Comment (danner26) [inline device-types/Juniper/EX4400-48P.yaml:120]: It looks like this devices power supplies are modular. If this is the case, then the PSUs should actually be `module-bays` and the applicable PSU units should be created and added to the module-types folder

File: device-types/Juniper/EX4400-48MP.yaml [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+    type: 100gbase-x-qsfp28
+  - name: et-0/1/1
+    type: 100gbase-x-qsfp28
+console-ports:
+  - name: Console
+    type: rj-45
+  - name: Front Console
+    type: usb-c
+power-ports:
+  - name: PSU0
```
Comment (danner26) [inline device-types/Juniper/EX4400-48MP.yaml:120]: It looks like this devices power supplies are modular. If this is the case, then the PSUs should actually be `module-bays` and the applicable PSU units should be created and added to the module-types folder

File: device-types/Juniper/EX4400-48F.yaml [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+    type: 100gbase-x-qsfp28
+  - name: et-0/1/1
+    type: 100gbase-x-qsfp28
+console-ports:
+  - name: Console
+    type: rj-45
+  - name: Front Console
+    type: usb-c
+power-ports:
+  - name: PSU0
```
Comment (danner26) [inline device-types/Juniper/EX4400-48F.yaml:119]: It looks like this devices power supplies are modular. If this is the case, then the PSUs should actually be `module-bays` and the applicable PSU units should be created and added to the module-types folder

File: device-types/Juniper/EX4400-24X.yaml [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+    type: 100gbase-x-qsfp28
+  - name: et-0/1/1
+    type: 100gbase-x-qsfp28
+console-ports:
+  - name: Console
+    type: rj-45
+  - name: Front Console
+    type: usb-c
+power-ports:
+  - name: PSU0
```
Comment (danner26) [inline device-types/Juniper/EX4400-24X.yaml:71]: It looks like this devices power supplies are modular. If this is the case, then the PSUs should actually be `module-bays` and the applicable PSU units should be created and added to the module-types folder

File: device-types/Juniper/EX4400-24T.yaml [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+    type: 100gbase-x-qsfp28
+  - name: et-0/1/1
+    type: 100gbase-x-qsfp28
+console-ports:
+  - name: Console
+    type: rj-45
+  - name: Front Console
+    type: usb-c
+power-ports:
+  - name: PSU0
```
Comment (danner26) [inline device-types/Juniper/EX4400-24T.yaml:71]: It looks like this devices power supplies are modular. If this is the case, then the PSUs should actually be `module-bays` and the applicable PSU units should be created and added to the module-types folder

File: device-types/Juniper/EX4400-24P.yaml [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+    type: 100gbase-x-qsfp28
+  - name: et-0/1/1
+    type: 100gbase-x-qsfp28
+console-ports:
+  - name: Console
+    type: rj-45
+  - name: Front Console
+    type: usb-c
+power-ports:
+  - name: PSU0
```
Comment (danner26) [inline device-types/Juniper/EX4400-24P.yaml:72]: It looks like this devices power supplies are modular. If this is the case, then the PSUs should actually be `module-bays` and the applicable PSU units should be created and added to the module-types folder

File: device-types/Juniper/EX4400-24MP.yaml [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+    type: 100gbase-x-qsfp28
+  - name: et-0/1/1
+    type: 100gbase-x-qsfp28
+console-ports:
+  - name: Console
+    type: rj-45
+  - name: Front Console
+    type: usb-c
+power-ports:
+  - name: PSU0
```
Comment (danner26) [inline device-types/Juniper/EX4400-24MP.yaml:72]: It looks like this devices power supplies are modular. If this is the case, then the PSUs should actually be `module-bays` and the applicable PSU units should be created and added to the module-types folder

File: device-types/Juniper/ACX7100-32C.yaml [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+  - name: et-0/0/32
+    type: 400gbase-x-qsfpdd
+  - name: et-0/0/33
+    type: 400gbase-x-qsfpdd
+  - name: et-0/0/34
+    type: 400gbase-x-qsfpdd
+  - name: et-0/0/35
+    type: 400gbase-x-qsfpdd
+module-bays:
+  - name: PSM 0
```
Comment (danner26) [inline device-types/Juniper/ACX7100-32C.yaml:87]: You may want to add a position definition to these, so if the PSU module type utilizes the standard `{module}` definition in the interface/port names they will evaluate properly

File: device-types/Juniper/ACX7100-32C.yaml [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+    # Comment: Power Supply Module 0
+    maximum_draw: 1600
+  - name: PSM 1
```
Comment (danner26) [inline device-types/Juniper/ACX7100-32C.yaml:90]: You may want to add a position definition to these, so if the PSU module type utilizes the standard `{module}` definition in the interface/port names they will evaluate properly

File: device-types/Juniper/ACX7100-32C.yaml [hunk 18]
```diff
@@ -0,0 +1,98 @@
+---
+manufacturer: Juniper
+model: ACX7100-32C
+slug: juniper-acx7100-32c
+u_height: 1
```
[the remark this answers is not in the document]
Comment (danner26) [reply device-types/Juniper/ACX7100-32C.yaml:5]: If the model or part number change based on the airflow, then yes there should be separate files for each

[the remark this answers is not in the document]
Comment (suom1) [reply device-types/Juniper/ACX7100-32C.yaml:5]: gotcha, I guess that applies for AC vs DC also in that case? Even when the only thing that changes in AC vs DC models are PSU's which are modular.

**Comment by suom1** [comment]
@danner26 could I get a new review? :)

**Comment by danner26** [comment]
Hey @suom1 sorry about that, I have been backed up lately. Working through these now

**Comment by suom1** [comment]
@danner26 Would it be possible to get this merged? :)

**Comment by danner26** [comment]
Hey @suom1 this is one of the much larger ones. I have limited time to work through all of the PRs, and the larger ones get pushed to the bottom. I will get it it ASAP though. Thanks for commenting so that it does not go stale

**Comment by suom1** [comment]
@danner26 is this still under review or is there something wrong with the PR?

**Comment by danner26** [comment]
Sorry I havent gotten around to this one since it is so large. I wil remove the stale flag and try to get to it ASAP

**Comment by danner26** [comment]
Hello, we have released a new set of tests and features to assist with properly creating definition files for the DeviceType-Library. These tests will ensure that all files conform to a standardized format and contain valid data. Over the coming weeks we will be adding more tests to handle other currently undefined cases. If you notice anything that you believe is a bug, please @ me in a comment and details what you think is incorrect.

Along with additional test cases we have added a feature that will annotate why tests fail. To access these annotations after a test has completed, please navigate to the _Files changed_ tab within your PR. At the very bottom of the page you will see a list of errors along with the filename in question and the error which was found.

With all this being said, if your PR is validated then great! Please @ me me so I can review it further! If your PR is not valid, please review the issues and correct them; in the meantime I will mark your PR as a draft. After correcting the issues and recieving a valid status, please mark the PR as ready to be reviewed. Thank you!

**Comment by suom1** [comment]
@danner26 ready once again for review!

**Review by danner26 — APPROVED** [review]

**Pull request merged**